contracts/serverless-v2/Jobs.sol

```diff
@@ -0,0 +1,502 @@
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable NO_OF_NODES_TO_SELECT;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable EXECUTOR_FEE_PER_MS;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable STAKING_REWARD_PER_MS;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    address public immutable STAKING_PAYMENT_POOL;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    address public immutable USDC_PAYMENT_POOL;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    Executors public immutable EXECUTORS;
+
+    //-------------------------------- Job start --------------------------------//
+
+    struct Job {
+        uint256 deadline;   // in milliseconds
+        uint256 execStartTime;
+        uint256 executionTime;   // it stores the execution time for first output submitted only (in milliseconds)
+        address jobOwner;
+        uint8 outputCount;
+        address[] selectedExecutors;
+        mapping(address => bool) hasExecutedJob;    // selectedExecutor => hasExecuted
+    }
+
+    // jobKey => Job
+    Job[] public jobs;
+
+    bytes32 private constant DOMAIN_SEPARATOR =
+        keccak256(
+            abi.encode(
+                keccak256("EIP712Domain(string name,string version)"),
+                keccak256("marlin.oyster.Jobs"),
+                keccak256("1")
+            )
+        );
+    bytes32 private constant SUBMIT_OUTPUT_TYPEHASH =
+        keccak256("SubmitOutput(uint256 jobId,bytes output,uint256 totalTime,uint8 errorCode,uint256 signTimestampInMs)");
+
+    event JobCreated(
+        uint256 indexed jobId,
+        address indexed jobOwner,
+        bytes32 codehash,
+        bytes codeInputs,
+        uint256 deadline,   // in milliseconds
+        address[] selectedExecutors
+    );
+
+    event JobResponded(
+        uint256 indexed jobId,
+        bytes output,
+        uint256 totalTime,
+        uint8 errorCode,
+        uint8 outputCount
+    );
+
+    event JobSucceeded(
+        uint256 indexed jobId,
+        bool callback_success
+    );
+
+    event JobTimeoutFailure(
+        uint256 indexed jobId,
+        bool callback_success
+    );
+
+    error JobsRelayTimeOver();
+    error JobsJobMarkedEndedAsResourceUnavailable();
+    error JobsInvalidSequenceId();
+    error JobsJobAlreadyRelayed();
+    error JobsUnsupportedChain();
+    error JobsSignatureTooOld();
+    error JobsExecutionTimeOver();
+    error JobsNotSelectedExecutor();
+    error JobsExecutorAlreadySubmittedOutput();
+
+    //-------------------------------- internal functions start --------------------------------//
+
+    function _createJob(
+        bytes32 _codehash,
+        bytes memory _codeInputs,
+        uint256 _deadline,  // in milliseconds
+        address _jobOwner
+    ) internal returns (uint256 jobId, uint8 errorCode) {
```

Comment:
try-catch might or might not be better for error handling like this

Reply:
you should be able to use `.selector` to get the error selector instead of calculating manually

Reply:
Done.